plat/marvell/a8k/common/plat_pm.c

```diff
@@ -0,0 +1,781 @@
+
+		/* trigger IPC message to MSS */
+		mss_pm_ipc_msg_trigger();
+
+		/* pm system synchronization */
+		bakery_lock_release(&pm_sys_lock);
+
+		/* trace message */
+		PM_TRACE(TRACE_PWR_DOMAIN_OFF);
+	} else {
+	INFO("%s: is not supported without SCP\n", __func__);
+	return;
+}
+}
+
+/* Get PM config to power off the SoC */
+void *plat_get_pm_cfg(void)
+{
+	return NULL;
+}
+
+/*
+ * This function should be called on restore from
+ * "suspend to RAM" state when the execution flow
+ * has to bypass BootROM image to RAM copy and speed up
+ * the system recovery
+ *
+ */
+static void plat_exit_bootrom(void)
+{
+	exit_bootrom(PLAT_MARVELL_TRUSTED_ROM_BASE);
+}
+
+/*
+ * Prepare for the power off of the system via GPIO
+ */
+static void plat_marvell_power_off_gpio(struct power_off_method *pm_cfg)
+{
+	unsigned int gpio;
+	unsigned int idx;
+	unsigned int shift;
+	unsigned int reg;
+	unsigned int addr;
+	gpio_info_t *info;
+	unsigned int tog_bits;
+
+	assert((pm_cfg->cfg.gpio.pin_count < PMIC_GPIO_MAX_NUMBER) &&
+	       (pm_cfg->cfg.gpio.step_count < PMIC_GPIO_MAX_TOGGLE_STEP));
+
+	/* Prepare GPIOs for PMIC */
+	for (gpio = 0; gpio < pm_cfg->cfg.gpio.pin_count; gpio++) {
+		info = &pm_cfg->cfg.gpio.info[gpio];
+		/* Set PMIC GPIO to output mode */
+		reg = mmio_read_32(MVEBU_CP_GPIO_DATA_OUT_EN(info->cp_index, info->gpio_index));
+		mmio_write_32(MVEBU_CP_GPIO_DATA_OUT_EN(info->cp_index, info->gpio_index),
+			      reg & ~MVEBU_GPIO_MASK(info->gpio_index));
+
+		/* Set the appropriate MPP to GPIO mode */
+		reg = mmio_read_32(MVEBU_PM_MPP_REGS(info->cp_index, info->gpio_index));
+		mmio_write_32(MVEBU_PM_MPP_REGS(info->cp_index, info->gpio_index),
+			reg & ~MVEBU_MPP_MASK(info->gpio_index));
+	}
+
+	/* Wait for MPP & GPIO pre-configurations done */
+	mdelay(pm_cfg->cfg.gpio.delay_ms);
+
+	/* Toggle the GPIO values, and leave final step to be triggered after  DDR self-refresh is enabled */
+	for (idx = 0; idx < pm_cfg->cfg.gpio.step_count; idx++) {
+		tog_bits = pm_cfg->cfg.gpio.seq[idx];
+
+		/* The GPIOs must be within same GPIO register, thus could get the original value by first GPIO */
+		info = &pm_cfg->cfg.gpio.info[0];
+		reg = mmio_read_32(MVEBU_CP_GPIO_DATA_OUT(info->cp_index, info->gpio_index));
+		addr = MVEBU_CP_GPIO_DATA_OUT(info->cp_index, info->gpio_index);
+
+		for (gpio = 0; gpio < pm_cfg->cfg.gpio.pin_count; gpio++) {
+			shift = pm_cfg->cfg.gpio.info[gpio].gpio_index % 32;
+			if (GPIO_LOW == (tog_bits & (1 << gpio)))
+				reg &= ~(1 << shift);
+			else
+				reg |= (1 << shift);
+		}
+
+		/* Set the GPIO register, for last step just store register address and values to system registers */
+		if (idx < pm_cfg->cfg.gpio.step_count - 1) {
+			mmio_write_32(MVEBU_CP_GPIO_DATA_OUT(info->cp_index, info->gpio_index), reg);
+			mdelay(pm_cfg->cfg.gpio.delay_ms);
+		} else {
+			/* Save GPIO register value to X17, and address to X18 */
+			__asm__ volatile (
+				"mov	x17, %0\n\t"
```

Comment:
How is this supposed to work? Won't the compiler simply remove it?

Reply:
Hmm... Agree, I should check it with the original code author and maybe look though the code disassembly.

Reply:
@antonio-nino-diaz-arm we are trying to keep these values in registers for ensuring there is no DRAM access for this data later. Any access to DRAM will return it from self-refresh mode. Do you have suggestion for reaching the same result without specifying register numbers here?

Reply:
The problem is that the compiler doesn't know that you want to keep the values in these registers. As soon as you exit this function, they can be used by the compiler. I see that you are using them in ``a8k_pwr_domain_pwr_down_wfi()``. I suspect that right now this works by pure luck because the code in between is relatively simple. Instead of doing this I suggest you to pass the values like this:
```
static void plat_marvell_power_off_gpio(struct power_off_method *pm_cfg, u_register_t *gpio_value, u_register_t *gpio_address)
{
    ...
    *gpio_value = reg;
    *gpio_address = addr;
    ...
}
```
And in ``a8k_pwr_domain_pwr_down_wfi()``:
```
u_register_t gpio_value, gpio_address;
plat_marvell_power_off_prepare(pm_cfg, &gpio_value, &gpio_address);
```
Then use ``gpio_value`` and ``gpio_address`` in your block of inline assembly.

The stack is in SRAM so DRAM accesses shouldn't be a problem with this solution.